mono/utils/mono-path.c

```diff
@@ -177,27 +177,129 @@ mono_path_resolve_symlinks (const char *path)
 #endif
 }
 
+#ifndef mono_host_win32
+#ifdef HOST_WIN32
+#define mono_host_win32 TRUE
+#else
+#define mono_host_win32 FALSE
+#endif
+#endif
+
+static gboolean
+mono_path_char_is_separator (char ch)
+{
+	return ch == '/' || (mono_host_win32 && ch == '\\');
+}
+
+static gboolean
+mono_path_contains_separator (const char *path, size_t length)
+{
+	for (size_t i = 0; i < length; ++i) {
+		if (mono_path_char_is_separator (path [i]))
+			return TRUE;
+	}
+	return FALSE;
+}
+
+static void
+mono_path_remove_trailing_path_separators (const char *path,  size_t *length)
+{
+	size_t i = *length;
+	while (i > 0 && mono_path_char_is_separator (path [i - 1]))
+		i -= 1;
+	*length = i;
+}
+
+static gboolean
+mono_path_char_is_lowercase (char ch)
+{
+	return mono_host_win32 && ch >= 'a' && ch <= 'z';
```

Comment:
File system case sensitivity is not locale sensitive. That would fail miserably. There is multi-user persistent sorted data structures. Locale sensitivity would break them. Upcase table is stored per volume by format & not the same through OS versions, but close.

There is an approximation in win32, but using 8 bit characters seems maybe far from correct in the first place. I can check.
CompareString with invariant culture or somesuch?

Yes we could depend on canonicalization wrt slashes.